module/os/freebsd/spl/spl_page.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+/* Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T */
+/* All Rights Reserved   */
+
+/*
+ * University Copyright- Copyright (c) 1982, 1986, 1988
+ * The Regents of the University of California
+ * All Rights Reserved
+ *
+ * University Acknowledgment- Portions of this document are derived from
+ * software developed by the University of California, Berkeley, and its
+ * contributors.
+ */
+
+/*
+ * $FreeBSD$
+ */
+
+#include <sys/page.h>
+#include <sys/byteorder.h>
+#include <sys/lock.h>
+#include <sys/vm.h>
+#include <vm/vm_map.h>
+#include <sys/zfs_context.h>
+
+/*
+ *
+ * if the operation is marked as read then we are stating the kernel will
+ * be writing to pages and must give it write access.
+ *
+ *
+ * This code comes from the FreeBSD's linuxkpi and models their implementation
+ * of get_user_pages(), set_page_dirty(), and put_page().
+ */
+long
+zfs_hold_pages(unsigned long start, unsigned long nr_pages, int read,
+    zfs_page_p *pages)
+{
+	vm_map_t map;
+	vm_prot_t prot;
+	size_t len;
+	int count;
+	int write = read ? 1 : 0;
+
+	map = &curthread->td_proc->p_vmspace->vm_map;
+
+	prot = write ? (VM_PROT_READ | VM_PROT_WRITE) : VM_PROT_READ;
+	len = ((size_t)nr_pages) << PAGE_SHIFT;
+	count = vm_fault_quick_hold_pages(map, start, len, prot, pages,
+	    nr_pages);
+
+	if (count == -1)
+		return (-EFAULT);
+	return (count);
+}
+
+long
+zfs_get_user_pages(unsigned long start, unsigned long nr_pages, int read,
+    zfs_page_p *pages)
+{
+	int count;
+
```

Comment:
I don't understand what you're trying to do here. Hold now wires the pages in head. So you're just doubling the work for no apparent reason.

Reply:
```
int
vm_fault_quick_hold_pages(vm_map_t map, vm_offset_t addr, vm_size_t len,
    vm_prot_t prot, vm_page_t *ma, int max_count)
{
	vm_offset_t end, va;
	vm_page_t *mp;
	int count;
	boolean_t pmap_failed;

	if (len == 0)
		return (0);
	end = round_page(addr + len);
	addr = trunc_page(addr);

	if (!vm_map_range_valid(map, addr, end))
		return (-1);

	if (atop(end - addr) > max_count)
		panic("vm_fault_quick_hold_pages: count > max_count");
	count = atop(end - addr);

	/*
	 * Most likely, the physical pages are resident in the pmap, so it is
	 * faster to try pmap_extract_and_hold() first.
	 */
	pmap_failed = FALSE;
	for (mp = ma, va = addr; va < end; mp++, va += PAGE_SIZE) {
		*mp = pmap_extract_and_hold(map->pmap, va, prot);
		if (*mp == NULL)

================

vm_page_t
pmap_extract_and_hold(pmap_t pmap, vm_offset_t va, vm_prot_t prot)
{
	pdp_entry_t pdpe, *pdpep;
	pd_entry_t pde, *pdep;
	pt_entry_t pte, PG_RW, PG_V;
	vm_page_t m;

	m = NULL;
	PG_RW = pmap_rw_bit(pmap);
	PG_V = pmap_valid_bit(pmap);
	PMAP_LOCK(pmap);
<snip>
	m = PHYS_TO_VM_PAGE(pte & PG_FRAME);

check_page:
	if (m != NULL && !vm_page_wire_mapped(m))
		m = NULL;
out:
	PMAP_UNLOCK(pmap);
	return (m);

```

Reply:
You have updated this in the current version that I just pushed.